implementations/rust/ockam/ockam_api/src/cloud/share/mod.rs

```diff
@@ -0,0 +1,102 @@
+use std::{fmt::Display, str::FromStr};
+
+use minicbor::{Decode, Encode};
+use serde::{Deserialize, Serialize};
+
+mod accept;
+mod create;
+mod list;
+
+pub use accept::*;
+pub use create::*;
+pub use list::*;
+
+#[derive(Clone, Debug, PartialEq, Decode, Deserialize, Encode, Serialize)]
+#[cbor(index_only)]
+#[rustfmt::skip]
+pub enum RoleInShare {
+    #[n(0)] Admin,
+    #[n(1)] Guest,
+    #[n(2)] Service,
+}
+
+impl Display for RoleInShare {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+        match self {
+            Self::Admin => write!(f, "admin"),
+            Self::Guest => write!(f, "guest"),
+            Self::Service => write!(f, "service_user"),
+        }
+    }
+}
+
+impl FromStr for RoleInShare {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "admin" => Ok(Self::Admin),
+            "guest" => Ok(Self::Guest),
+            "service" => Ok(Self::Service),
+            other => Err(format!("unknown role: {other}")),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Decode, Deserialize, Encode, Serialize)]
+#[cbor(index_only)]
+#[rustfmt::skip]
+pub enum ShareScope {
+    #[n(0)] Project,
```

Comment:
Are we supposed to share a full project / space? What does that mean? Does it mean that any relay created on a project will be accessible then? If you could collect the data model in a small diagram we could discuss that during the next pod meeting (either to get agreement or to present it to others)

Reply:
Also I don't understand the interactions between `ShareScope` and `target_id`. What is the difference between a `CreateInvite` with a `Service` scope and a `CreateServiceInvite`?

Reply:
Repeating from another comment since it's contextual here too:
> The Service invite needs additional mandatory information to resolve, so it's special cased in a few spots.

A `CreateInvite` with Service scope doesn't contain all the needed details and would always fail on the request to the backend, so I should probably make a constructor function that provides that feedback more immediately without needing to reach a backend. It's valid to be in any of the three scopes in the read responses, which is why I've currently got it in the enum. I'll ponder whether any other redesigns would be better for safety/correctness or otherwise help prevent misuse.

Reply:
Yes, if it's not complicated I think it'd be better to rule out cases that would fail.

Reply:
I believe this is also captured in spirit as part of #5517, please let me know if that's not the case.